include/deal.II/base/mpi_consensus_algorithms.templates.h

```diff
@@ -207,14 +190,36 @@ namespace Utilities
       NBX<T1, T2>::all_locally_originated_receives_are_completed()
       {
 #ifdef DEAL_II_WITH_MPI
-        int        all_receive_requests_are_done;
-        const auto ierr = MPI_Testall(recv_requests.size(),
-                                      recv_requests.data(),
-                                      &all_receive_requests_are_done,
-                                      MPI_STATUSES_IGNORE);
-        AssertThrowMPI(ierr);
+        // We know that all requests have come in when we have pending
+        // messages from all targets with the right tag. We can check
+        // for pending messages with MPI_IProbe, which returns
+        // immediately with a return code that indicates whether
+        // it has found a message from a given process with a given
+        // tag
+        for (const unsigned int target : targets)
+          {
+            const int tag_deliver = Utilities::MPI::internal::Tags::
+              consensus_algorithm_nbx_process_deliver;
+
+            int        request_is_pending;
+            const auto ierr = MPI_Iprobe(target,
+                                         tag_deliver,
+                                         this->comm,
+                                         &request_is_pending,
+                                         MPI_STATUS_IGNORE);
+            AssertThrowMPI(ierr);
+
+            // If there is no pending message from that process,
+            // then we are clearly not done receiving everything
+            // yet -- so return false:
+            if (request_is_pending == false)
```

Comment:
I think this is an integer-to-bool comparison here. I wish MPI would take bools, but instead we have to write
```suggestion
            if (request_is_pending == 0)
```